onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -0,0 +1,481 @@
+
+  inline void MergePrediction1(OTYPE* /*predictions*/, unsigned char* /*has_predictions*/,
+                               OTYPE* /*predictions2*/, unsigned char* /*has_predictions2*/) const {}
+
+  inline void FinalizeScores1(OTYPE* Z, OTYPE& val,
+                              unsigned char& has_scores,
+                              int64_t*) const {
+    val = has_scores ? (val + origin_) : origin_;
+    *Z = this->post_transform_ == POST_EVAL_TRANSFORM::PROBIT ? static_cast<OTYPE>(ComputeProbit(static_cast<float>(val))) : val;
+  }
+
+  // N outputs
+
+  void ProcessTreeNodePrediction(OTYPE* /*predictions*/, TreeNodeElement<OTYPE>* /*rooOTYPE*/,
+                                 unsigned char* /*has_predictions*/) const {}
+
+  void MergePrediction(int64_t /*n*/, OTYPE* /*predictions*/, unsigned char* /*has_predictions*/,
+                       OTYPE* /*predictions2*/, unsigned char* /*has_predictions2*/) const {}
+
+  void FinalizeScores(std::vector<OTYPE>& scores,
+                      std::vector<unsigned char>& has_scores,
+                      OTYPE* Z, int add_second_class,
+                      int64_t*) const {
+    OTYPE val;
+    for (int64_t jt = 0; jt < n_targets_or_classes_; ++jt) {
+      val = use_base_values_ ? (*base_values_)[jt] : 0.f;
+      val += has_scores[jt] ? scores[jt] : 0;
+      scores[jt] = val;
+    }
+    write_scores(scores, post_transform_, Z, add_second_class);
+  }
+};
+
+/////////////
+// regression
+/////////////
+
+template <typename ITYPE, typename OTYPE>
+class _AggregatorSum : public _Aggregator<ITYPE, OTYPE> {
+ public:
+  inline _AggregatorSum(size_t n_trees,
+                        const int64_t& n_targets_or_classes,
+                        POST_EVAL_TRANSFORM post_transform,
+                        const std::vector<OTYPE>* base_values) : _Aggregator<ITYPE, OTYPE>(n_trees, n_targets_or_classes,
+                                                                                           post_transform, base_values) {}
+
+  // 1 output
+
+  inline void ProcessTreeNodePrediction1(OTYPE* predictions,
+                                         TreeNodeElement<OTYPE>* root,
+                                         unsigned char*) const {
+    *predictions += root->weights[0].value;
+  }
+
+  inline void MergePrediction1(OTYPE* predictions, unsigned char*,
+                               const OTYPE* predictions2, const unsigned char*) const {
+    *predictions += *predictions2;
+  }
+
+  inline void FinalizeScores1(OTYPE* Z, OTYPE& val,
+                              unsigned char&,
+                              int64_t*) const {
+    val += this->origin_;
+    *Z = this->post_transform_ == POST_EVAL_TRANSFORM::PROBIT ? static_cast<OTYPE>(ComputeProbit(static_cast<float>(val))) : val;
+  }
+
+  // N outputs
+
+  void ProcessTreeNodePrediction(OTYPE* predictions, TreeNodeElement<OTYPE>* root,
+                                 unsigned char* has_predictions) const {
```

Comment:
This seems potentially error-prone as 'predictions' is passed in as a raw pointer and there's nothing to be able to validate that the length of that buffer matches values in root->weights.i.
I'd suggest gsl::span<OTYPE> for predictions would be safer as it can do the bounds checking in a debug build (turned off in a release build for performance reasons though).

Reply:
Fixed.